python/pyspark/sql/functions.pyi

```diff
@@ -359,13 +360,13 @@ def variance(col: ColumnOrName) -> Column: ...
 @overload
 def udf(
     f: Callable[..., Any], returnType: DataTypeOrString = ...
-) -> Callable[..., Column]: ...
+) -> UserDefinedFunction: ...
```

Comment:
It returns a wrapped function. So the previous signature is correct.

Reply:
`asNondeterministic` etc. is manually set to the specific wrapped function. Hm, we should fix the issue though

Reply:
@zero323 do you have any thought on this?

Reply:
Yeah, this should be addressed, but I am not sure what is the best approach here. We'll definitely need a protocol  ‒ at first glance something like this should be sufficient:

```python
class UserDefinedFunctionWrapper(Protocol):
    def __call__(self, *args: Column) -> Column: ...
    def asNonDeterministic(self) -> UserDefinedFunctionWrapper: ...
```
and shouldn't break any existing annotations, but I haven't tested it thoroughly. 

@luranhe Thanks for your proposal. Would you mind creating a [JIRA ticket](https://issues.apache.org/jira/projects/SPARK/issues/) and updating the title accordingly (it should be of form `[SPARK-XXXXX][PYTHON] ...`). Thanks in advance.

Reply:
Yeah, let's do that

Reply:
Also, `UserDefinedFunctionWrapper` is just a typing helper without corresponding class, so let's move it into `pyspark.sql._typing.pyi`.

Reply:
Just took a look, should we be using `UserDefinedFunctionLike` [here](https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/python/pyspark/sql/_typing.pyi#L56)?

Reply:
Yes, nice catch @luranhe!